src/protocol/DataFeed.sol

```diff
@@ -27,33 +27,34 @@ contract DataFeed is IDataFeed {
         publicationHashes.push(0);
     }
 
-    /// @notice Publish arbitrary data in blobs for data availability.
-    /// @param numBlobs the number of blobs accompanying this function call.
-    /// @param queries the calls required to retrieve L1 metadata hashes associated with this publication.
-    /// @dev there can be multiple queries because a single publication might represent multiple rollups,
-    /// each with their own L1 metadata requirements
-    /// @dev append a hash representing all blobs and L1 metadata to `publicationHashes`.
-    /// The number of blobs is not validated. Additional blobs are ignored. Empty blobs have a hash of zero.
-    function publish(uint256 numBlobs, MetadataQuery[] calldata queries) external payable onlyStandaloneTx {
+    /// @inheritdoc IDataFeed
+    function publish(uint256 numBlobs, bytes calldata data, MetadataQuery[] calldata queries)
+        external
+        payable
+        onlyStandaloneTx
+    {
         require(numBlobs > 0, "no data to publish");
 
         uint256 nQueries = queries.length;
         uint256 id = publicationHashes.length;
+
+        bytes32[] memory blobHashes = new bytes32[](numBlobs);
+        for (uint256 i; i < numBlobs; ++i) {
+            blobHashes[i] = blobhash(i);
+        }
+        bytes32 dataHash = keccak256(abi.encode(data, blobHashes));
+
         Publication memory publication = Publication({
             id: id,
             prevHash: publicationHashes[id - 1],
             publisher: msg.sender,
             timestamp: block.timestamp,
             blockNumber: block.number,
-            blobHashes: new bytes32[](numBlobs),
+            dataHash: dataHash,
```

Comment:
do we want to emit the `data` as part of the event? If not, and when `publish` is called in wrapping transaction, it's really hard to extract the `data` from the inner transaction -- even if we use use geth's tracing API.

Reply:
If later we remove the following:

```
bytes32[] memory blobHashes = new bytes32[](numBlobs);
        for (uint256 i; i < numBlobs; ++i) {
            blobHashes[i] = blobhash(i);
        }
```

We can remove the `uint256 numBlobs` input in the `publish` function, then maybe we can simply emit the `data` as part of the publication, not its hash.